### PR TITLE
/pin-clusters with redis (plus /heatmap)

### LINKED ISSUE #543: [FEAT] Implement a redis cache layer
### Overview
We have access to a free version of redis on heroku, may as well take advantage of it
See #450 for details

### Action Items
  - [ ] Create redis docker-compose entry in both the actual compose file and the example compose file
  - [ ] Establish cache-key hierarchy (Probably hashed versions of the filter params)
    - ojf84u9r8h44gh9eg -> hashed version of something like (highland park, past 6mo, bulky items)
    - ojf84u9r8h44gh9eg:pins:all
    - ojf84u9r8h44gh9eg:pins:{srnumber}
    - ojf84u9r8h44gh9eg:ttc
    - ojf84u9r8h44gh9eg:freq
  - [ ] Enforce a global TTL for all keys
  - [ ] Ensure connections are short lived
  - [ ] Make redis optional
  - [ ] Use LFU for key recycling

### Resources/Instructions
The caching algorithm will operate similar to this:
User selects filters
front end sends requests to backend
upon receiving request, hash the input parameters
```
Check redis for...
- ojf84u9r8h44gh9eg:pins:all
- ojf84u9r8h44gh9eg:ttc
- ojf84u9r8h44gh9eg:freq
```
If the key is found, retrieve values and return to front end

If nothing is found, continue down normal paths for responses and update the redis keys with the global TTL for 
```
- ojf84u9r8h44gh9eg:pins:all
- ojf84u9r8h44gh9eg:ttc
- ojf84u9r8h44gh9eg:freq
```
then return values to front end

In summary they cache will operate as either a GET then RETURN or a GET, NOTFOUND, SET, RETURN

### PR DESCRIPTION
**Pull request by jmensch1**
Fixes #543 and #565 (the backend part)

This creates the `/pin-clusters` endpoint with redis support. 

I ran some jmeter tests comparing the existing `/pins` endpoint with the new `/pin-clusters` endpoint. For the test I ran 3 separate api calls against each of the two endpoints:

**small**: 1 NC
**medium**: 10 NCs
**large**: 50 NCs

All of the api calls used a 1-year date range (2019-01-01 through 2020-01-01) and asked for all 11 request types.

I disabled redis during the test, and proxied my localhost through an `ngrok` url to simulate live-server conditions. 

These were the results:

### pins (small) vs pin-clusters (small)
pins-comparison-small

### pins (medium) vs pin-clusters (medium)
pins-comparison-medium

### pins (large) vs pin-clusters (large)
pins-comparison-large

So basically, `/pin-clusters` is faster than `/pins` for all filters, and the advantage grows as the filters get broader. This makes sense, since the payload for `/pin-clusters` stays essentially the same size, while the payload for `/pins` grows with the breadth of the filters.

Obvi `/pin-clusters` is still way too slow for medium and large requests. I'm pretty sure that's related to the inefficiency of IN queries (see #499) as well as the amount of time it takes to serialize rows from the database. I'm gonna look at that.

UPDATE:
Added a `/heatmap` endpoint that takes the standard filters and returns an array of lat/longs to support the heatmap. It returns the same data as the `/pins` endpoint, except it doesn't include the `srnumber` and `requesttype`. Also it uses the redis cache, and uses the same keys as the `/pin-clusters` endpoint because it requires a subset of that data. 

  - [x] Up to date with `dev` branch
  - [x] Branch name follows [guidelines](https://github.com/hackforla/311-data/blob/master/GETTING_STARTED.md#feature-branching)
  - [x] All PR Status checks are successful
  - [ ] Peer reviewed and approved

Any questions? See the [getting started guide](https://github.com/hackforla/311-data/blob/master/GETTING_STARTED.md)

### CODE REVIEW

File: server/src/settings.example.cfg [hunk 1]
```diff
@@ -37,3 +37,7 @@ GITHUB_TOKEN = REDACTED
 ISSUES_URL = https://api.github.com/repos/hackforla/311-data/issues
 PROJECT_URL = REDACTED
 GITHUB_SHA = DEVELOPMENT
+
+[Redis]
+ENABLED = True
```
Comment (sellnat77) [inline server/src/settings.example.cfg:42]: Can we default this to false?
Also is there a url for redis? Im guessing its defaulting to localhost:6379
In prod we will have a environment variable override for this

OOORRRRR we can base `ENABLED` off of 'was there a redis url provided or not' that way we kill two birds with 1 stone

**Comment by sellnat77** [comment]
These perf results are awesome, gonna be a game changer for sure :rocket:

File: server/src/settings.example.cfg (hunk 1 above)
Comment (jmensch1) [reply server/src/settings.example.cfg:42]: yeah I can change that. 

for the url, right now it's using 'redis' as the host, which I guess is the hostname of the redis service within the docker compose network...if I understand docker correctly. And the port is 6379. I'll figure out how to set this up as an env variable

Comment (sellnat77) [reply server/src/settings.example.cfg:42]: ohhh okay yeah...makes sense
Within the docker env itll be exposed as both redis:6379 and localhost:6379 so either way we want to be explicit with how we address redis so we can override it with heroku's `fn8934fh3o8g3o3qhg893gg.heroku.com:6379`

Comment (jmensch1) [reply server/src/settings.example.cfg]: ok cool this is done

**Pull request merged**